gst/nnstreamer/tensor_query/tensor_query_client.c

```diff
@@ -122,5 +302,430 @@ static void
+  GSocketAddress *saddr = NULL;
+  GSocket *socket = NULL;
+  GError *err = NULL;
+
+  if (is_sink) {
+    host = self->sink_host;
+    port = self->sink_port;
+    cancellable = self->sink_cancellable;
+  } else {
+    host = self->src_host;
+    port = self->src_port;
+    cancellable = self->src_cancellable;
+  }
+  socket = gst_tensor_query_socket_new (host, port, cancellable, &saddr);
+  if (!socket) {
+    nns_loge ("Failed to create new socket or connect to host");
+    g_object_unref (saddr);
+    return FALSE;
+  }
+
+  if (!g_socket_connect (socket, saddr, cancellable, &err)) {
+    if (g_error_matches (err, G_IO_ERROR, G_IO_ERROR_CANCELLED)) {
+      nns_logd ("Cancelled connecting");
+    } else {
+      nns_loge ("Failed to connect to host");
+    }
+    g_clear_error (&err);
+    g_object_unref (saddr);
+    return FALSE;
+  }
+  g_object_unref (saddr);
+
+  if (is_sink) {
+    self->sink_socket = socket;
+    GST_OBJECT_FLAG_SET (self, GST_TENSOR_QUERY_CLIENT_SINK_SOCKET_OPEN);
+  } else {
+    self->src_socket = socket;
+    GST_OBJECT_FLAG_SET (self, GST_TENSOR_QUERY_CLIENT_SRC_SOCKET_OPEN);
+  }
+  return TRUE;
+}
+
+/**
+ * @brief Called when the element starts processing. optional vmethod of BaseTransform
+ * @param[in] trans "this" pointer
+ * @return TRUE if there is no error.
+ */
+static gboolean
+gst_tensor_query_client_start (GstBaseTransform * trans)
+{
+  GstTensorQueryClient *self = GST_TENSOR_QUERY_CLIENT (trans);
+
+  /** Connect to tensor_query_src */
+  if (!GST_OBJECT_FLAG_IS_SET (self, GST_TENSOR_QUERY_CLIENT_SINK_SOCKET_OPEN)) {
+    if (!gst_tensor_query_client_connect (self, TRUE)) {
+      nns_loge ("Failed to connect tensor_query_src");
+      return FALSE;
+    }
+  }
+  /** Connect to tensor_query_sink */
+  if (!GST_OBJECT_FLAG_IS_SET (self, GST_TENSOR_QUERY_CLIENT_SRC_SOCKET_OPEN)) {
+    if (!gst_tensor_query_client_connect (self, FALSE)) {
+      nns_loge ("Failed to connect tensor_query_sink");
+      return FALSE;
+    }
+  }
+
+  return TRUE;
+}
+
+/**
+ * @brief Close the socket
+ */
+static gboolean
+gst_tensor_query_client_close_socket (GstTensorQueryClient * self,
+    GSocket ** socket)
+{
+  GError *err = NULL;
+
+  GST_DEBUG_OBJECT (self, "closing socket");
+  if (!g_socket_close (*socket, &err)) {
+    GST_ERROR_OBJECT (self, "Failed to close socket: %s", err->message);
+    g_clear_error (&err);
+    return FALSE;
+  }
+  g_object_unref (*socket);
+  *socket = NULL;
+
+  return TRUE;
+}
+
+/**
+ * @brief Called when the element stops processing. optional vmethod of BaseTransform
+ * @param[in] trans "this" pointer
+ * @return TRUE if there is no error.
+ */
+static gboolean
+gst_tensor_query_client_stop (GstBaseTransform * trans)
+{
+  GstTensorQueryClient *self = GST_TENSOR_QUERY_CLIENT (trans);
+
+  if (self->sink_socket) {
+    if (!gst_tensor_query_client_close_socket (self, &self->sink_socket)) {
+      nns_loge ("Failed to close sink socket");
+      return FALSE;
```

Comment:
IMO, even if ```gst_tensor_query_client_close_socket``` fails,  ```GST_OBJECT_FLAG_UNSET (self, GST_TENSOR_QUERY_CLIENT_SINK_SOCKET_OPEN);``` looks required. Is it possible to re-invoke ```gst_tensor_query_client_start``` later?

Reply:
> IMO, even if gst_tensor_query_client_close_socket fails, GST_OBJECT_FLAG_UNSET (self, GST_TENSOR_QUERY_CLIENT_SINK_SOCKET_OPEN); looks required.

Agree. Updated.

> Is it possible to re-invoke gst_tensor_query_client_start later?

`gst_tensor_query_client_start ` called when pad become activated.. 
Currently, it will only be called when state is changed.